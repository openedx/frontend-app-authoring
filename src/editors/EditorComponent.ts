/** Shared interface that all Editor components (like ProblemEditor) adhere to */
export interface EditorComponent {
  onClose: (() => void) | null;
  // TODO: get a better type for the 'result' here
  returnFunction?: (() => (result: any) => void) | null;
}
