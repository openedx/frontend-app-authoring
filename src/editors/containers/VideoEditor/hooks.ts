import { useState, createContext } from 'react';
import { thunkActions } from '../../data/redux';

import { StrictDict } from '../../utils';

type ErrRecord = Record<string, string>;
type ErrCategory = [errors: ErrRecord, setter: (newErrors: ErrRecord) => void];
interface ErrorContextData {
  duration: ErrCategory;
  handout: ErrCategory;
  license: ErrCategory;
  thumbnail: ErrCategory;
  transcripts: ErrCategory;
  videoSource: ErrCategory;
}

/* istanbul ignore next */
export const ErrorContext = createContext<ErrorContextData>({
  duration: [{}, () => {}],
  handout: [{}, () => {}],
  license: [{}, () => {}],
  thumbnail: [{}, () => {}],
  transcripts: [{}, () => {}],
  videoSource: [{}, () => {}],
});

export const state = StrictDict({
  /* eslint-disable react-hooks/rules-of-hooks */
  durationErrors: (val) => useState(val),
  handoutErrors: (val) => useState(val),
  licenseErrors: (val) => useState(val),
  thumbnailErrors: (val) => useState(val),
  transcriptsErrors: (val) => useState(val),
  videoSourceErrors: (val) => useState(val),
  /* eslint-enable react-hooks/rules-of-hooks */
});

export const errorsHook = (): { error: ErrorContextData, validateEntry: () => boolean } => {
  const [durationErrors, setDurationErrors] = state.durationErrors({});
  const [handoutErrors, setHandoutErrors] = state.handoutErrors({});
  const [licenseErrors, setLicenseErrors] = state.licenseErrors({});
  const [thumbnailErrors, setThumbnailErrors] = state.thumbnailErrors({});
  const [transcriptsErrors, setTranscriptsErrors] = state.transcriptsErrors({});
  const [videoSourceErrors, setVideoSourceErrors] = state.videoSourceErrors({});

  return {
    error: {
      duration: [durationErrors, setDurationErrors],
      handout: [handoutErrors, setHandoutErrors],
      license: [licenseErrors, setLicenseErrors],
      thumbnail: [thumbnailErrors, setThumbnailErrors],
      transcripts: [transcriptsErrors, setTranscriptsErrors],
      videoSource: [videoSourceErrors, setVideoSourceErrors],
    },
    validateEntry: () => {
      if (Object.keys(durationErrors).length > 0) { return false; }
      if (Object.keys(handoutErrors).length > 0) { return false; }
      if (Object.keys(licenseErrors).length > 0) { return false; }
      if (Object.keys(thumbnailErrors).length > 0) { return false; }
      if (Object.keys(transcriptsErrors).length > 0) { return false; }
      if (Object.keys(videoSourceErrors).length > 0) { return false; }
      return true;
    },
  };
};

export const fetchVideoContent = () => ({ dispatch }) => (
  dispatch(thunkActions.video.saveVideoData())
);
