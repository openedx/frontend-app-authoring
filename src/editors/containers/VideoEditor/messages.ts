import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  spinnerScreenReaderText: {
    id: 'authoring.videoEditor.spinnerScreenReaderText',
    defaultMessage: 'loading',
    description: 'Loading message for spinner screenreader text.',
  },
  replaceVideoButtonLabel: {
    id: 'authoring.videoEditor.replaceVideoButtonLabel',
    defaultMessage: 'Replace video',
    description: 'Text of the replace video button to return to the video gallery',
  },
});

export default messages;
