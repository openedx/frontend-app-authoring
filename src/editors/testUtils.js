/* istanbul ignore file */
import react from 'react';
import { StrictDict } from './utils';

/**
 * Mocked formatMessage provided by react-intl
 */
export const formatMessage = (msg, values) => {
  let message = msg.defaultMessage;
  if (values === undefined) {
    return message;
  }
  Object.keys(values).forEach((key) => {
    // eslint-disable-next-line
    message = message.replace(`{${key}}`, values[key]);
  });
  return message;
};

/**
 * Mock a single component, or a nested component so that its children render nicely
 * in snapshots.
 * @param {string} name - parent component name
 * @param {obj} contents - object of child components with intended component
 *   render name.
 * @return {func} - mock component with nested children.
 *
 * usage:
 *   mockNestedComponent('Card', { Body: 'Card.Body', Form: { Control: { Feedback: 'Form.Control.Feedback' }}... });
 *   mockNestedComponent('IconButton', 'IconButton');
 */
export const mockNestedComponent = (name, contents) => {
  if (typeof contents !== 'object') {
    return contents;
  }
  const fn = () => name;
  Object.defineProperty(fn, 'name', { value: name });
  Object.keys(contents).forEach((nestedName) => {
    const value = contents[nestedName];
    fn[nestedName] = typeof value !== 'object'
      ? value
      : mockNestedComponent(`${name}.${nestedName}`, value);
  });
  return fn;
};

/**
 * Mock a module of components.  nested components will be rendered nicely in snapshots.
 * @param {obj} mapping - component module mock config.
 * @return {obj} - module of flat and nested components that will render nicely in snapshots.
 * usage:
 *   mockNestedComponents({
 *     Card: { Body: 'Card.Body' },
 *     IconButton: 'IconButton',
 *   })
 */
export const mockNestedComponents = (mapping) => Object.entries(mapping).reduce(
  (obj, [name, value]) => ({
    ...obj,
    [name]: mockNestedComponent(name, value),
  }),
  {},
);

/**
 * Mock utility for working with useState in a hooks module.
 * Expects/requires an object containing the state object in order to ensure
 * the mock behavior works appropriately.
 *
 * Expected format:
 *   hooks = { state: { <key>: (val) => React.createRef(val), ... } }
 *
 * Returns a utility for mocking useState and providing access to specific state values
 * and setState methods, as well as allowing per-test configuration of useState value returns.
 *
 * Example usage:
 *   // hooks.js
 *   import * as module from './hooks';
 *   const state = {
 *     isOpen: (val) => React.useState(val),
 *     hasDoors: (val) => React.useState(val),
 *     selected: (val) => React.useState(val),
 *   };
 *   ...
 *   export const exampleHook = () => {
 *     const [isOpen, setIsOpen] = module.state.isOpen(false);
 *     if (!isOpen) { return null; }
 *     return { isOpen, setIsOpen };
 *   }
 *   ...
 *
 *   // hooks.test.js
 *   import * as hooks from './hooks';
 *   const state = new MockUseState(hooks)
 *   ...
 *   describe('state hooks', () => {
 *     state.testGetter(state.keys.isOpen);
 *     state.testGetter(state.keys.hasDoors);
 *     state.testGetter(state.keys.selected);
 *   });
 *   describe('exampleHook', () => {
 *     beforeEach(() => { state.mock(); });
 *     it('returns null if isOpen is default value', () => {
 *       expect(hooks.exampleHook()).toEqual(null);
 *     });
 *     it('returns isOpen and setIsOpen if isOpen is not null', () => {
 *       state.mockVal(state.keys.isOpen, true);
 *       expect(hooks.exampleHook()).toEqual({
 *         isOpen: true,
 *         setIsOpen: state.setState[state.keys.isOpen],
 *       });
 *     });
 *     afterEach(() => { state.restore(); });
 *   });
 *
 * @param {obj} hooks - hooks module containing a 'state' object
 */
export class MockUseState {
  constructor(hooks) {
    this.hooks = hooks;
    this.oldState = null;
    this.setState = {};
    this.stateVals = {};

    this.mock = this.mock.bind(this);
    this.restore = this.restore.bind(this);
    this.mockVal = this.mockVal.bind(this);
    this.testGetter = this.testGetter.bind(this);
  }

  /**
   * @return {object} - StrictDict of state object keys
   */
  get keys() {
    return StrictDict(Object.keys(this.hooks.state).reduce(
      (obj, key) => ({ ...obj, [key]: key }),
      {},
    ));
  }

  /**
   * Replace the hook module's state object with a mocked version, initialized to default values.
   */
  mock() {
    this.oldState = this.hooks.state;
    Object.keys(this.keys).forEach(key => {
      this.hooks.state[key] = jest.fn(val => {
        this.stateVals[key] = val;
        return [val, this.setState[key]];
      });
    });
    this.setState = Object.keys(this.keys).reduce(
      (obj, key) => ({
        ...obj,
        [key]: jest.fn(val => {
          this.hooks.state[key] = val;
        }),
      }),
      {},
    );
  }

  /**
   * Restore the hook module's state object to the actual code.
   */
  restore() {
    this.hooks.state = this.oldState;
  }

  /**
   * Mock the state getter associated with a single key to return a specific value one time.
   * @param {string} key - state key (from this.keys)
   * @param {any} val - new value to be returned by the useState call.
   */
  mockVal(key, val) {
    this.hooks.state[key].mockReturnValueOnce([val, this.setState[key]]);
  }

  testGetter(key) {
    test(`${key} state getter should return useState passthrough`, () => {
      const testValue = 'some value';
      const useState = (val) => ({ useState: val });
      jest.spyOn(react, 'useState').mockImplementationOnce(useState);
      expect(this.hooks.state[key](testValue)).toEqual(useState(testValue));
    });
  }
}
