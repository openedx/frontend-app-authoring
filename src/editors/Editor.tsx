// Note: there is no Editor.test.tsx. This component only works together with
// <EditorPage> as its parent, so they are tested together in EditorPage.test.tsx
import React from 'react';
import { useDispatch } from 'react-redux';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import messages from './messages';
import * as hooks from './hooks';

import supportedEditors from './supportedEditors';
import type { EditorComponent } from './EditorComponent';
import { useEditorContext } from './EditorContext';

export interface Props extends EditorComponent {
  blockType: string;
  blockId: string | null;
  learningContextId: string | null;
  lmsEndpointUrl: string | null;
  studioEndpointUrl: string | null;
  fullScreen?: boolean;
}

const Editor: React.FC<Props> = ({
  learningContextId,
  blockType,
  blockId,
  lmsEndpointUrl,
  studioEndpointUrl,
  onClose = null,
  returnFunction = null,
}) => {
  const dispatch = useDispatch();
  hooks.initializeApp({
    dispatch,
    data: {
      blockId,
      blockType,
      learningContextId,
      lmsEndpointUrl,
      studioEndpointUrl,
    },
  });
  const { fullScreen } = useEditorContext();

  const EditorComponent = supportedEditors[blockType];
  const innerEditor = (EditorComponent !== undefined)
    ? <EditorComponent {...{ onClose, returnFunction }} />
    : <FormattedMessage {...messages.couldNotFindEditor} />;

  if (fullScreen) {
    return (
      <div
        className="d-flex flex-column"
      >
        <div
          className="pgn__modal-fullscreen h-100"
          role="dialog"
          aria-label={blockType}
        >
          {innerEditor}
        </div>
      </div>
    );
  }
  return innerEditor;
};

export default Editor;
