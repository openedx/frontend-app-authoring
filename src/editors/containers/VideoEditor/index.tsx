import React from 'react';
import { useSelector } from 'react-redux';
import {
  Spinner,
} from '@openedx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { selectors } from '../../data/redux';
import { RequestKeys } from '../../data/constants/requests';

import { EditorComponent } from '../../EditorComponent';
import EditorContainer from '../EditorContainer';
import VideoEditorModal from './components/VideoEditorModal';
import { ErrorContext, errorsHook, fetchVideoContent } from './hooks';
import messages from './messages';

const VideoEditor: React.FC<EditorComponent> = ({
  onClose,
  returnFunction,
}) => {
  const intl = useIntl();
  const studioViewFinished = useSelector(
    (state) => selectors.requests.isFinished(state, { requestKey: RequestKeys.fetchStudioView }),
  );
  const isLibrary = useSelector(selectors.app.isLibrary) as boolean;
  const {
    error,
    validateEntry,
  } = errorsHook();
  return (
    <ErrorContext.Provider value={error}>
      <EditorContainer
        getContent={fetchVideoContent()}
        isDirty={/* istanbul ignore next */ () => true}
        onClose={onClose}
        returnFunction={returnFunction}
        validateEntry={validateEntry}
      >
        {studioViewFinished ? (
          <div className="video-editor">
            <VideoEditorModal {...{ isLibrary }} />
          </div>
        ) : (
          <div style={{
            position: 'absolute',
            left: '50%',
            top: '50%',
            transform: 'translate(-50%, -50%)',
          }}
          >
            <Spinner
              animation="border"
              className="m-3"
              screenreadertext={intl.formatMessage(messages.spinnerScreenReaderText)}
            />
          </div>
        )}
      </EditorContainer>
    </ErrorContext.Provider>
  );
};

export default VideoEditor;
