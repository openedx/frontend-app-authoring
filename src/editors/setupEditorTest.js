// These additional mocks and setup are required for some tests in src/editors/
// and are imported on an as-needed basis.

// /////////////////////////////////////////////////////////////////////////////
// TODO: tests using this 'setupEditorTest', shallow rendering, and snapshots
// should be replaced with modern tests that use src/testUtils.ts and
// @testing-library/react. See
//   src/editors/containers/ProblemEditor/components/SelectTypeModal/index.test.tsx
// for an example of an editor test in the new format.
// /////////////////////////////////////////////////////////////////////////////

import { formatMessage as mockFormatMessage } from './testUtils';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-canvas-mock';

jest.mock('@edx/frontend-platform/i18n', () => {
  const i18n = jest.requireActual('@edx/frontend-platform/i18n');
  const PropTypes = jest.requireActual('prop-types');
  return {
    ...i18n,
    useIntl: () => ({ formatMessage: mockFormatMessage }),
    intlShape: PropTypes.shape({
      formatMessage: PropTypes.func,
    }),
    defineMessages: m => m,
    getLocale: () => 'getLocale',
    FormattedDate: () => 'FormattedDate',
    FormattedMessage: () => 'FormattedMessage',
    FormattedTime: () => 'FormattedTime',
  };
});

jest.mock('@openedx/paragon', () => jest.requireActual('./testUtils').mockNestedComponents({
  Alert: {
    Heading: 'Alert.Heading',
  },
  AlertModal: 'AlertModal',
  ActionRow: {
    Spacer: 'ActionRow.Spacer',
  },
  Badge: 'Badge',
  Button: 'Button',
  ButtonGroup: 'ButtonGroup',
  Collapsible: {
    Advanced: 'Advanced',
    Body: 'Body',
    Trigger: 'Trigger',
    Visible: 'Visible',
  },
  Card: {
    Header: 'Card.Header',
    Section: 'Card.Section',
    Footer: 'Card.Footer',
    Body: 'Card.Body',
  },
  CheckboxControl: 'CheckboxControl',
  Col: 'Col',
  Container: 'Container',
  Dropdown: {
    Item: 'Dropdown.Item',
    Menu: 'Dropdown.Menu',
    Toggle: 'Dropdown.Toggle',
  },
  ErrorContext: {
    Provider: 'ErrorContext.Provider',
  },
  Hyperlink: 'Hyperlink',
  Icon: 'Icon',
  IconButton: 'IconButton',
  IconButtonWithTooltip: 'IconButtonWithTooltip',
  Image: 'Image',
  MailtoLink: 'MailtoLink',
  ModalDialog: {
    Footer: 'ModalDialog.Footer',
    Header: 'ModalDialog.Header',
    Title: 'ModalDialog.Title',
    Body: 'ModalDialog.Body',
    CloseButton: 'ModalDialog.CloseButton',
  },
  Form: {
    Checkbox: 'Form.Checkbox',
    Control: {
      Feedback: 'Form.Control.Feedback',
    },
    Group: 'Form.Group',
    Label: 'Form.Label',
    Text: 'Form.Text',
    Row: 'Form.Row',
    Radio: 'Radio',
    RadioSet: 'RadioSet',
  },
  OverlayTrigger: 'OverlayTrigger',
  Tooltip: 'Tooltip',
  FullscreenModal: 'FullscreenModal',
  Row: 'Row',
  Scrollable: 'Scrollable',
  SelectableBox: {
    Set: 'SelectableBox.Set',
  },

  Spinner: 'Spinner',
  Stack: 'Stack',
  Toast: 'Toast',
  Truncate: 'Truncate',
  useWindowSize: { height: '500px' },
}));

jest.mock('@openedx/paragon/icons', () => ({
  Close: jest.fn().mockName('icons.Close'),
  Edit: jest.fn().mockName('icons.Edit'),
  Locked: jest.fn().mockName('icons.Locked'),
  Unlocked: jest.fn().mockName('icons.Unlocked'),
}));

// Mock react-redux hooks
// unmock for integration tests
jest.mock('react-redux', () => {
  const dispatch = jest.fn((...args) => ({ dispatch: args })).mockName('react-redux.dispatch');
  return {
    connect: (mapStateToProps, mapDispatchToProps) => (component) => ({
      mapStateToProps,
      mapDispatchToProps,
      component,
    }),
    useDispatch: jest.fn(() => dispatch),
    useSelector: jest.fn((selector) => ({ useSelector: selector })),
  };
});

// Mock the plugins repo so jest will stop complaining about ES6 syntax
jest.mock('frontend-components-tinymce-advanced-plugins', () => ({
  a11ycheckerCss: '',
}));
