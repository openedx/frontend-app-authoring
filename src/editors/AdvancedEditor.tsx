import React, { useEffect } from 'react';
import { getConfig } from '@edx/frontend-platform';
import { useIntl } from '@edx/frontend-platform/i18n';

import { LibraryBlock } from '../library-authoring/LibraryBlock';
import { EditorModalWrapper } from './containers/EditorContainer';
import { ToastContext } from '../generic/toast-context';
import messages from './messages';

interface AdvancedEditorProps {
  usageKey: string,
  onClose: Function | null,
}

const AdvancedEditor = ({ usageKey, onClose }: AdvancedEditorProps) => {
  const intl = useIntl();
  const { showToast } = React.useContext(ToastContext);

  useEffect(() => {
    const handleIframeMessage = (event) => {
      if (event.origin !== getConfig().STUDIO_BASE_URL) {
        return;
      }

      if (event.data.type === 'xblock-event') {
        const { eventName, data } = event.data;

        if (onClose && (eventName === 'cancel'
          || (eventName === 'save' && data.state === 'end'))
        ) {
          onClose();
        } else if (eventName === 'error') {
          showToast(intl.formatMessage(messages.advancedEditorGenericError));
        }
      }
    };

    window.addEventListener('message', handleIframeMessage);

    return () => {
      window.removeEventListener('message', handleIframeMessage);
    };
  }, []);

  return (
    <EditorModalWrapper onClose={onClose as () => void}>
      <LibraryBlock
        usageKey={usageKey}
        view="studio_view"
      />
    </EditorModalWrapper>
  );
};

export default AdvancedEditor;
