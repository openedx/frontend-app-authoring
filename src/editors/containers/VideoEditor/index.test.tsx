import 'CourseAuthoring/editors/setupEditorTest';
import { shallow } from '@edx/react-unit-test-utils';

import { formatMessage } from '../../testUtils';
import VideoEditor from '.';

jest.mock('../EditorContainer', () => 'EditorContainer');
jest.mock('./components/VideoEditorModal', () => 'VideoEditorModal');

jest.mock('./hooks', () => ({
  ErrorContext: {
    Provider: 'ErrorContext.Provider',
  },
  errorsHook: jest.fn(() => ({
    error: 'hooks.errorsHook.error',
    validateEntry: jest.fn().mockName('validateEntry'),
  })),
  fetchVideoContent: jest.fn().mockName('fetchVideoContent'),
}));

jest.mock('../../data/redux', () => ({
  selectors: {
    requests: {
      isFinished: jest.fn((state, params) => ({ isFailed: { state, params } })),
    },
    app: {
      isLibrary: jest.fn(state => ({ isLibrary: state })),
    },
  },
}));

jest.mock('@openedx/paragon', () => ({
  ...jest.requireActual('@openedx/paragon'),
  Spinner: 'Spinner',
}));

describe('VideoEditor', () => {
  const props = {
    onClose: jest.fn().mockName('props.onClose'),
    intl: { formatMessage },
    studioViewFinished: false,
    isLibrary: false,
  };
  describe('snapshots', () => {
    test('renders as expected with default behavior', () => {
      expect(shallow(<VideoEditor {...props} />).snapshot).toMatchSnapshot();
    });
  });
});
