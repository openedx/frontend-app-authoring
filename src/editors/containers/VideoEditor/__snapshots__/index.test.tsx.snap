// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoEditor snapshots renders as expected with default behavior 1`] = `
<ErrorContext.Provider
  value="hooks.errorsHook.error"
>
  <EditorContainer
    isDirty={[Function]}
    onClose={[MockFunction props.onClose]}
    validateEntry={[MockFunction validateEntry]}
  >
    <div
      className="video-editor"
    >
      <VideoEditorModal
        isLibrary={
          {
            "useSelector": [MockFunction],
          }
        }
      />
    </div>
  </EditorContainer>
</ErrorContext.Provider>
`;
